### ISSUE TITLE
Since 1.5.1. switch from URLConnectionEngine to ApacheHttpClient43Engine impacts in very poor client performance

### ISSUE THREAD

**Issue by banovp** [quarkusio/quarkus#10160]
**Describe the bug**
We have performance problems with a Restservice Client based on _resteasy-client-microprofile 4.5.3._ caused by the forced usage of the **ApacheHttpClient43Engine**. Please note, the worst performance is the same, regardless of the tls usage (**both http and https** have the same miserable performance)

Within the 1.5.1.Final there was the issue:
https://github.com/quarkusio/quarkus/pull/9773

which might cause the problem (we assume this, cause this issue is described as the one forcing the usage of ApacheHttpClient43Engine). Currently each http call needs a 1,5sec, previously the same request to the same mock service took up to 50ms. Now when we replace the _org.jboss.resteasy.microprofile.client.RestClientBuilderImpl_ class and force the usage of the URLConnectionEngine it's again very very fast also with underlying new Quarkus 1.5.1.Final. or 1.5.2.Final

Our code - RestService init within a Singleton
````
import org.eclipse.microprofile.rest.client.RestClientBuilder;
....

@Singleton
public class RestServiceAccess {

    @Produces
    public RestService getProxy() {
        try {
            return RestClientBuilder.newBuilder()
                    .baseUrl(new URL(url))
                    .trustStore(********)
                    .build(RestService.class);
        } catch (MalformedURLException e) {
        ......
        }
    }
````

usage of the produced client within the singleton:

```
@Singleton
public class AService {

    @Inject
    RestService restService;
    
    public void something(){
        restService.aMethod();
    }

```

the RestService
```
/**
 * Interface. Just used to build our REST client!
 */
@Produces("application/xml")
@Consumes("application/xml")
public interface RestService {
    @POST
    @Path("/aMethod")
    @Retry(maxRetries = 1, delay = 1, delayUnit = ChronoUnit.SECONDS, maxDuration = 20, durationUnit = ChronoUnit.SECONDS)
    @Timeout(10000L)
    Replay aMethod(Request request);
```

Right now as a workaround we are overwriting the **RestClientBuilderImpl** and forcing the usage of the **URLConnectionEngine** to have the same performance as before the upgrade. Please let us know, if the usage of the **ApacheHttpClient43Engine** is causing some new instancing of the clients on each call or something similar, which could cause this big performance issue. 

tx

P.S.
our dirty fix for now is basically:

replacing line 265 in RestClientBuilderImpl#build and forcing always the usage of the **URLConnectionClientEngineBuilder** for any case.
```
      //Always using the URLConnectionEngine
        resteasyClientBuilder.httpEngine(new URLConnectionClientEngineBuilder().resteasyClientBuilder(resteasyClientBuilder).build());

        if (!SSL_ENABLED) {
            resteasyClientBuilder.sslContext(null);
            resteasyClientBuilder.trustStore(null);
            resteasyClientBuilder.keyStore(null, "");
        }
```

this results in the same performance as before.

**Comment by geoand** [comment]
@kenfinnigan @asoldano have you encountered this sort of problem before?

**Comment by kenfinnigan** [comment]
Are there things you're doing with `RestClientBuilder` that precludes you from using direct CDI injection of the rest-client instance? Not an issue, more a curiosity.

This is for a test situation right, as you mention the mock taking 50ms previously?

Is it possible that the rest-client call is triggering retries and or timeouts which is why the response is taking longer?

**Comment by banovp** [comment]
Hi, 
we are using a mock service only for testing and after deployment the client is calling a remote service. In both cases a client based on _ApacheHttpClient43Engine_ is very very slow, there is no real difference in the performance, it's just very bad. With the URLConnectionEngine on the other hand both clients are very fast, as before. To force usage of the URLConnectionEngine we've replaced the RestClientBuilderImpl (as described above)

We are using the _RestClientBuilder_ to set a custom truststore (see in code).

````
return RestClientBuilder.newBuilder()
                    .baseUrl(new URL(url))
                    .trustStore(********)
`````

In the log of the remote system or in log of the mock service, which is called by the _ApacheHttpClient43Engine_ based clients, no retries are visible, also in our log we are not able to see a retry in the log. With debuging is also no retry noticeable. It's just very slow by a normal http or https call. That's why the question above, are you aware of new instance creation within the request processing when the _ApacheHttpClient43Engine_ is used, or some further configuration is needed?

tx,
best regards

**Comment by jaikiran** [comment]
Hello @banovp, is there an application you can share with us (preferably as a github repo) which reproduces this issue? Or maybe your application.properties contents?

**Comment by gsmet** [comment]
@banovp maybe you could profile what's going on using this guide: https://github.com/quarkusio/quarkus/blob/master/TROUBLESHOOTING.md ?

**Comment by banovp** [comment]
Hi all, 

Thank you for your effort. As we were building a demo project today for you to reproduce the issue, we were able to solve the problem. It was our mock service. The mock service was not closing the connection:
```
    private static class RequestHandler implements HttpHandler {

        @Override
        public void handleRequest(HttpServerExchange exchange) throws Exception {
            exchange.setStatusCode(200);
            exchange.addResponseHeader("Content-Type", "application/xml");
            Response response = new Response();
            response.setElement2("ok");
            final byte[] xmlReply = XMLUtil.getInstance().getMarshalledXML(response, XMLUtil.XMLFormat.PRETTY_PRINT);
            exchange.writeBlocking(Unpooled.wrappedBuffer(xmlReply), true);
        }
```

we've added 
```
exchange.addResponseHeader("Connection", "close");
```

and this solved the issue. The **ApacheHttpClient43Engine** is using connection pooling. This was not the case previously with the **URLConnectionClient**. So each time a connection was not closed, the next call crashed by taking this connection from the pool. Our service was annotated with:

```
@Retry(maxRetries = 1, delay = 1, delayUnit = ChronoUnit.SECONDS, maxDuration = 20, durationUnit = ChronoUnit.SECONDS)
```
and was retrying automatically, without a hint in the log(my bad), so the next call was happy, but the complete new init of the client was taking so much time, because the connection in the pool was crashed by the previous call.  So finally each request was crashing and then the retry was successful, which was not visible in the logs. 

In the same time as we found the issue we've had most probably network issues and our remote calls to a real service based tests were also very slow.... Now the remote calls are fast and apparently both issue are completely independent. After the fix on the mock service our integration testing is running smoothly. So basically this was the issue.

Thank you again for your help!

Best Regards
